File: apps/ios/Barns/BarnsApp/Presentation/Patterns/PatternDetailViewModel.swift
import Combine

@MainActor
final class PatternDetailViewModel: ObservableObject {
    enum State: Equatable {
        case loading
        case loaded(WallGreenPattern)
        case notFound
        case failed(String)
    }

    @Published private(set) var state: State = .loading

    private let patternId: WallGreenPattern.ID
    private let getPatternDetailUseCase: GetPatternDetailUseCase

    init(patternId: WallGreenPattern.ID, getPatternDetailUseCase: GetPatternDetailUseCase) {
        self.patternId = patternId
        self.getPatternDetailUseCase = getPatternDetailUseCase
    }

    func load() async {
        state = .loading
        do {
            if let pattern = try await getPatternDetailUseCase.execute(id: patternId) {
                state = .loaded(pattern)
            } else {
                state = .notFound
            }
        } catch {
            state = .failed("Unable to load this pattern. Please try again.")
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/patterns/PatternDetailViewModel.kt
package com.barns.app.presentation.patterns

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.WallGreenPattern
import com.barns.app.domain.usecase.patterns.GetPatternDetailUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class PatternDetailViewModel(
    private val patternId: String,
    private val getPatternDetailUseCase: GetPatternDetailUseCase,
) : ViewModel() {
    sealed interface State {
        object Loading : State
        data class Loaded(val pattern: WallGreenPattern) : State
        object NotFound : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    fun load() {
        _state.value = State.Loading
        viewModelScope.launch {
            runCatching { getPatternDetailUseCase.execute(patternId) }
                .onSuccess { pattern ->
                    _state.value = if (pattern != null) State.Loaded(pattern) else State.NotFound
                }
                .onFailure { _state.value = State.Failed("Unable to load this pattern. Please try again.") }
        }
    }
}
